core/blockchain.go

```diff
@@ -3015,3 +3034,155 @@ func EnablePersistDiff(limit uint64) BlockChainOption {
 		return chain
 	}
 }
+
+func (bc *BlockChain) GetRootByDiffHash(blockNumber uint64, blockHash common.Hash, diffHash common.Hash) (*types.VerifyResult, error) {
+	var res types.VerifyResult
+	res.BlockNumber = blockNumber
+	res.BlockHash = blockHash
+
+	if blockNumber > bc.CurrentHeader().Number.Uint64()+11 {
+		res.Status = types.StatusBlockTooNew
+		return &res, nil
+	} else if blockNumber > bc.CurrentHeader().Number.Uint64() {
+		res.Status = types.StatusBlockNewer
+		return &res, nil
+	}
+
+	diff := bc.GetTrustedDiffLayer(blockHash)
+	if diff != nil {
+		if diff.DiffHash == (common.Hash{}) {
+			hash, err := GetTrustedDiffHash(diff)
+			if err != nil {
+				res.Status = types.StatusUnexpectedError
+				return &res, err
+			}
+
+			diff.DiffHash = hash
```

Comment:
there is a race condition on `diff.DiffHash`,  concurrently read and write on `DiffHash`. Try using `atomic` to avoid the race condition.

Reply:
replace by atomic.value